### PR TITLE
Set read buffer to 0

### PR DESCRIPTION
**Pull request by cboden**
refs reactphp/event-loop#26

### CODE REVIEW

**Comment by cboden** [comment]
Courtesy of @mbonneau

**Comment by cboden** [comment]
I'd like to merge this by June 8th unless any valid objections. ping @clue @WyriHaximus

File: tests/StreamTest.php [hunk 1]
```diff
@@ -113,6 +113,40 @@ public function testClosingStreamInDataEventShouldNotTriggerError()
         $conn->handleData($stream);
     }
 
+    public function testBufferReadsLargeChunks() {
```
Comment (clue) [inline tests/StreamTest.php]: All other methods test the `Stream` in isolation by mocking the environment (unit tests), while this test uses the environment (integration test). Perhaps move this to a separate test file? *(Also, there's a missing newline)*

File: tests/StreamTest.php [hunk 2]
```diff
@@ -113,6 +113,40 @@ public function testClosingStreamInDataEventShouldNotTriggerError()
...
+        $loop = \React\EventLoop\Factory::create();
```
Comment (clue) [inline tests/StreamTest.php]: This only picks the "best" available event loop implementation…
IMO we should run this test against all available event loop implementations (data provider?).

File: tests/StreamTest.php [hunk 3]
```diff
@@ -113,6 +113,40 @@ public function testClosingStreamInDataEventShouldNotTriggerError()
...
+        if ($loop instanceof \React\EventLoop\StreamSelectLoop) {
+            return $this->markTestSkipped('Regression bug only works on extension loops');
```
Comment (clue) [inline tests/StreamTest.php]: IMO the same test should also run (and work) with the `StreamSelectLoop`.

**Comment by clue** [comment]
The code looks good to me and I'm all for getting this in! :+1: We're already handling our own buffers, so it makes sense to use unbuffered I/O on a lower level.

Given how long it took to trace this (subtle) bug, we should take the time and make sure to not introduce any regressions. In particular:

* How does this affect file I/O?
* How does this affect TCP server sockets (socket component)?
* How does this affect TCP clients (socket-client component)?
* How does this affect SSL/TLS clients (socket-client component)?
* Performance regressions?

File: src/Stream.php [hunk 4]
```diff
@@ -24,6 +24,7 @@ public function __construct($stream, LoopInterface $loop)
         }
 
         stream_set_blocking($this->stream, 0);
+        stream_set_read_buffer($this->stream, 0);
```
Comment (clue) [inline src/Stream.php]: So far this only affects *reading* from the stream – how about also disabling the *write* buffer?

Comment (cboden) [inline src/Stream.php]: I'm leaning towards "if it ain't broke don't fix it". The read buffer is fixing a specific problem. I'm all for trying to set the write buffer to 0 in another PR if we can show it increasing performance or fixing a bug.

Comment (clue) [inline src/Stream.php]: > if it ain't broke don't fix it

Sounds fair :)

**Comment by KingCrunch** [comment]
@cboden What's missing for a merge?

**Comment by WyriHaximus** [comment]
LGTM :+1:

**Comment by mbonneau** [comment]
+1

**Comment by davidwdan** [comment]
:+1:

**Comment by clue** [comment]
> Given how long it took to trace this (subtle) bug, we should take the time and make sure to not introduce any regressions […]

We have yet to spot any regressions, so I see no point in holding this off any longer. Let's get this in :+1:

**Comment by clue** [comment]
Note that this PR introduced an incompatibility with HHVM which is being addressed in #28.

**Pull request merged**